README.md
---
author: mfw78 <mfw78@protonmail.com>
---

# WindingTree DAO

## Overview

This repository contains the smart contracts deployed for Governance use by WindingTree DAO.

Currently the following tools are deployed:

- [`TimelockController`](https://etherscan.io/address/0xaeb7b8808ce9afc9730846ec81880b57658734dc)

## TimelockController

The `TimelockController` is designed to provide a _time-delay_ mechanism for Governance Functions.
For information relating to the administration and configuration of the `TimelockController`,
please refer to the [OpenZeppelin documentation](https://docs.openzeppelin.com/contracts/4.x/api/governance#TimelockController).

Current configuration (may be verified on chain):

- `DEFAULT_ADMIN_ROLE`: N/A
- `TIMELOCK_ADMIN_ROLE`: `0xAEB7b8808ce9afc9730846Ec81880B57658734dC` (self-administered)
- `PROPOSAL_ROLE`: `0x876969b13dcf884C13D4b4f003B69229E6b7966A` (community multi-sig)
- `EXECUTOR_ROLE`: `0x0000000000000000000000000000000000000000` (Permissionless execution)
- `minDelay`: 259200 (seconds, ie. 3 days)


CHANGELOG.md
## 0.0.1 (2022-04-04)

### Bug Fixes

- 🐛 license ([7322695](https://github.com/windingtree/smartcontracts-hardhat-template/commit/73226950b883921cd6deaef4259b9f4cccfa9c31))

### Features

- 🎸 docgen ([cc0f0de](https://github.com/windingtree/smartcontracts-hardhat-template/commit/cc0f0de354a27bf0cbd97c9b3739781c85626c1d))
- 🎸 smart contract template repo ([adad158](https://github.com/windingtree/smartcontracts-hardhat-template/commit/adad15812f33610e877f1d8740ed5a177e90d2aa))


test/utils/index.ts
import { Contract } from 'ethers'
import { ethers } from 'hardhat'

/**
 * A helper function that assists with connecting to contracts from an EOA.
 * @param addresses array of user addresses to connect to applicable contracts
 * @param contracts contracts that are being connected to from a user
 * @returns an array of users with their connected contracts
 */
export async function setupUsers<T extends { [contractName: string]: Contract }>(
  addresses: string[],
  contracts: T
): Promise<({ address: string } & T)[]> {
  const users: ({ address: string } & T)[] = []
  for (const address of addresses) {
    users.push(await setupUser(address, contracts))
  }
  return users
}

/**
 * A helper function to connect to contracts from a specific user.
 * @param address the address used to connect to the contracts
 * @param contracts contracts to connect to
 * @returns an object of connected contracts
 */
export async function setupUser<T extends { [contractName: string]: Contract }>(
  address: string,
  contracts: T
): Promise<{ address: string } & T> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const user: any = { address }
  for (const key of Object.keys(contracts)) {
    user[key] = contracts[key].connect(await ethers.getSigner(address))
  }
  return user as { address: string } & T
}


test/MockERC20.spec.ts
import { ethers, getNamedAccounts, deployments, getUnnamedAccounts } from 'hardhat'

import { setupUser, setupUsers } from './utils'

import { IERC20Metadata as IERC20 } from '../typechain'
import { expect } from './chai-setup'
import { utils } from 'ethers'

const setup = deployments.createFixture(async () => {
  await deployments.fixture('MockERC20')
  const { deployer, alice, bob, carol } = await getNamedAccounts()
  const contracts = {
    erc20: (await ethers.getContract('MockERC20')) as IERC20
  }
  const users = await setupUsers(await getUnnamedAccounts(), contracts)

  return {
    users,
    deployer: await setupUser(deployer, contracts),
    alice: await setupUser(alice, contracts),
    bob: await setupUser(bob, contracts),
    carol: await setupUser(carol, contracts),
    ...contracts
  }
})

describe('MockERC20', function () {
  // let users: ({ address: string } & { erc20: IERC20 })[]
  let alice: { address: string } & { erc20: IERC20 }
  let bob: { address: string } & { erc20: IERC20 }
  let carol: { address: string } & { erc20: IERC20 }

  beforeEach('load fixture', async () => {
    // eslint-disable-next-line @typescript-eslint/no-extra-semi
    ;({ alice, bob, carol } = await setup())
  })

  context('Metadata', async () => {
    it('sets symbol correctly', async () => {
      expect(await alice.erc20.symbol()).to.be.eq('MTK')
      expect(await alice.erc20.name()).to.be.eq('MockERC20')
    })
  })

  context('Allocations', async () => {
    it('gives correct amount to alice', async () => {
      expect(await alice.erc20.balanceOf(alice.address)).to.be.eq(utils.parseEther('1000000'))
    })
    it('gives correct amount to bob', async () => {
      expect(await bob.erc20.balanceOf(alice.address)).to.be.eq(utils.parseEther('1000000'))
    })
    it('gives correct amount to carol', async () => {
      expect(await carol.erc20.balanceOf(alice.address)).to.be.eq(utils.parseEther('1000000'))
    })
  })
})


test/TimelockController.ts
import { ethers, getNamedAccounts, deployments, getUnnamedAccounts } from 'hardhat'

import { setupUser, setupUsers } from './utils'

import { expect } from './chai-setup'
import { constants } from 'ethers'
import { TimelockController } from '../typechain'

const COMMUNITY_MULTI_SIG = '0x876969b13dcf884C13D4b4f003B69229E6b7966A'
const MIN_DELAY = 60 * 60 * 24 * 3 // 3 days

const DEFAULT_ADMIN_ROLE = '0x0000000000000000000000000000000000000000000000000000000000000000'
const TIMELOCK_ADMIN_ROLE = '0x5f58e3a2316349923ce3780f8d587db2d72378aed66a8261c916544fa6846ca5'
const PROPOSER_ROLE = '0xb09aa5aeb3702cfd50b6b62bc4532604938f21248a27a1d5ca736082b6819cc1'
const EXECUTOR_ROLE = '0xd8aa0f3194971a2a116679f7c2090f6939c8d4e01a2a8d7e41d55e5351469e63'

const setup = deployments.createFixture(async () => {
  await deployments.fixture('TimelockController')
  const { deployer, alice, bob, carol } = await getNamedAccounts()
  const contracts = {
    timelockController: (await ethers.getContract('TimelockController')) as TimelockController
  }
  const users = await setupUsers(await getUnnamedAccounts(), contracts)

  return {
    users,
    deployer: await setupUser(deployer, contracts),
    alice: await setupUser(alice, contracts),
    bob: await setupUser(bob, contracts),
    carol: await setupUser(carol, contracts),
    ...contracts
  }
})

describe('TimelockController', function () {
  let deployer: { address: string } & { timelockController: TimelockController }
  let alice: { address: string } & { timelockController: TimelockController }

  beforeEach('load fixture', async () => {
    // eslint-disable-next-line @typescript-eslint/no-extra-semi
    ;({ deployer, alice } = await setup())
  })

  context('access controls set correctly', async () => {
    it('deployer does not have any roles', async () => {
      expect(await alice.timelockController.hasRole(DEFAULT_ADMIN_ROLE, deployer.address)).to.be.equal(false)
      expect(await alice.timelockController.hasRole(TIMELOCK_ADMIN_ROLE, deployer.address)).to.be.equal(false)
      expect(await alice.timelockController.hasRole(PROPOSER_ROLE, deployer.address)).to.be.equal(false)
      expect(await alice.timelockController.hasRole(EXECUTOR_ROLE, deployer.address)).to.be.equal(false)
    })
    it('timelock controller is itself a timelock admin', async () => {
      expect(await alice.timelockController.hasRole(TIMELOCK_ADMIN_ROLE, alice.timelockController.address)).to.be.equal(
        true
      )
    })
    it('community multi-sig is NOT a timelock admin', async () => {
      expect(await alice.timelockController.hasRole(TIMELOCK_ADMIN_ROLE, COMMUNITY_MULTI_SIG)).to.be.equal(false)
    })
    it('community multi-sig is a proposer', async () => {
      expect(await alice.timelockController.hasRole(PROPOSER_ROLE, COMMUNITY_MULTI_SIG)).to.be.equal(true)
    })
    it('any address is an executor', async () => {
      expect(await alice.timelockController.hasRole(EXECUTOR_ROLE, constants.AddressZero)).to.be.equal(true)
    })
  })

  context('parameters set correctly', async () => {
    it('set minimum delay to 3 days', async () => {
      expect(await alice.timelockController.getMinDelay()).to.be.equal(MIN_DELAY)
    })
  })
})


hardhat.config.ts
import { HardhatUserConfig } from 'hardhat/types'

import '@openzeppelin/hardhat-upgrades'
import '@nomiclabs/hardhat-ethers'
import '@nomiclabs/hardhat-waffle'
import '@nomiclabs/hardhat-solhint'
import '@typechain/hardhat'
import 'hardhat-deploy'
import 'hardhat-gas-reporter'
import 'solidity-coverage'
import 'solidity-docgen'
import { nodeUrl, accounts } from './utils/network'

import 'dotenv/config'

const config: HardhatUserConfig = {
  solidity: {
    compilers: [
      {
        version: '0.8.13',
        settings: {
          optimizer: {
            enabled: true,
            runs: 200
          }
        }
      }
    ]
  },
  namedAccounts: {
    deployer: 0,
    alice: 1,
    bob: 2,
    carol: 3
  },
  networks: {
    hardhat: {
      // process.env.HARDHAT_FORK will specify the network that the fork is made from.
      // this line ensure the use of the corresponding accounts
      accounts: accounts(process.env.HARDHAT_FORK),
      hardfork: 'london',
      // gasPrice: 11000000000,
      forking: process.env.HARDHAT_FORK
        ? {
            // TODO once PR merged : network: process.env.HARDHAT_FORK,
            url: nodeUrl(process.env.HARDHAT_FORK),
            blockNumber: process.env.HARDHAT_FORK_NUMBER ? parseInt(process.env.HARDHAT_FORK_NUMBER) : undefined
          }
        : undefined,
      saveDeployments: false,
      tags: ['local', 'forked']
    },
    ganache: {
      url: 'http://127.0.0.1:7545/',
      saveDeployments: true,
      tags: ['local']
    },
    localhost: {
      url: nodeUrl('localhost'),
      accounts: accounts(),
      saveDeployments: true,
      tags: ['local']
    },
    staging: {
      url: nodeUrl('rinkeby'),
      accounts: accounts('rinkeby'),
      tags: ['staging']
    },
    production: {
      url: nodeUrl('mainnet'),
      accounts: accounts('mainnet'),
      tags: ['production']
    },
    mainnet: {
      url: nodeUrl('mainnet'),
      accounts: accounts('mainnet'),
      tags: ['production']
    },
    rinkeby: {
      url: nodeUrl('rinkeby'),
      accounts: accounts('rinkeby'),
      tags: ['staging']
    },
    kovan: {
      url: nodeUrl('kovan'),
      accounts: accounts('kovan'),
      tags: ['staging']
    },
    goerli: {
      url: nodeUrl('goerli'),
      accounts: accounts('goerli'),
      tags: ['staging']
    },
    sokol: {
      url: nodeUrl('sokol'),
      accounts: accounts('sokol'),
      tags: ['staging']
    },
    gnosis: {
      url: nodeUrl('gnosis'),
      accounts: accounts('gnosis'),
      tags: ['production']
    }
  },
  gasReporter: {
    currency: 'USD',
    gasPrice: 100,
    enabled: !!process.env.REPORT_GAS,
    coinmarketcap: process.env.COINMARKETCAP_API_KEY,
    maxMethodDiff: 10
  },
  typechain: {
    outDir: 'typechain',
    target: 'ethers-v5'
  },
  mocha: {
    timeout: 0
  },
  verify: {
    etherscan: {
      apiKey: process.env.ETHERSCAN_API_KEY
    }
  }
}

export default config


deploy/001.ts
/* eslint-disable camelcase */
import { HardhatRuntimeEnvironment } from 'hardhat/types'
import { DeployFunction } from 'hardhat-deploy/types'
import { ethers } from 'hardhat'
import { constants } from 'ethers'
import { TimelockController } from '../typechain'

const COMMUNITY_MULTI_SIG = '0x876969b13dcf884C13D4b4f003B69229E6b7966A'
const MIN_DELAY = 60 * 60 * 24 * 3 // 3 days

const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
  const { deployments, getNamedAccounts } = hre
  const { deploy } = deployments

  const { deployer, alice, bob, carol } = await getNamedAccounts()

  // --- Account listing ---
  console.log(`Deployer: ${deployer}`)
  console.log(`Alice: ${alice}`)
  console.log(`Bob: ${bob}`)
  console.log(`Carol: ${carol}`)

  // --- Deploy the contract
  const timelockControllerDeploy = await deploy('TimelockController', {
    from: deployer,
    log: true,
    autoMine: true,
    args: [MIN_DELAY, [COMMUNITY_MULTI_SIG], [constants.AddressZero]] // minimum delay, community multi-sig as only proposer and any executor
  })

  if (timelockControllerDeploy.newlyDeployed) {
    console.log(
      `TimelockController deployed at ${timelockControllerDeploy.address} using ${timelockControllerDeploy.receipt?.gasUsed} gas`
    )

    const timelockControllerFactory = await ethers.getContractFactory('TimelockController')
    const timelockController = timelockControllerFactory.attach(timelockControllerDeploy.address) as TimelockController

    // Set timelockcontroller as timelockcontroller admin.
    const tx = await timelockController.renounceRole(
      await timelockController.callStatic.TIMELOCK_ADMIN_ROLE(),
      deployer
    )
    const receipt = await tx.wait()

    console.log(receipt)
  }
}

export default func
func.tags = ['TimelockController']
